lib/bundler/diff.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module Bundler
+  class Diff
+    def initialize(string1, string2)
+      @string1 = string1
+      @string2 = string2
+    end
+
+    def diff
+      @diff ||= begin
+          @paths = [tempfile(@string1), tempfile(@string2)]
+
+          cmd = format '"%s" %s %s', diff_bin, "-U10000", @paths.map {|s| %("#{s}") }.join(" ")
+          diff = `#{cmd}`
+          diff.force_encoding("ASCII-8BIT") if diff.respond_to?(:valid_encoding?) && !diff.valid_encoding?
+          diff
+        end
+    ensure
+      # unlink the tempfiles explicitly now that the diff is generated
+      if defined? @tempfiles # to avoid Ruby warnings about undefined ins var.
+        Array(@tempfiles).each do |t|
+          begin
+            # check that the path is not nil and file still exists.
+            # REE seems to be very agressive with when it magically removes
+            # tempfiles
+            t.unlink if t.path && File.exist?(t.path)
+          rescue StandardError => e
+            warn "#{e.class}: #{e}"
+            warn e.backtrace.join("\n")
+          end
+        end
+      end
+    end
+
+    def to_s
+      diff
+      regexp = /^(--- "?#{@paths[0]}"?|\+\+\+ "?#{@paths[1]}"?|@@|\\\\)/
+      @diff.split("\n").reject {|x| x =~ regexp }.map {|line| line + "\n" }.join
+    end
+
+  private
+
+    def tempfile(string)
+      t = Tempfile.new("diffy")
+      # ensure tempfiles aren't unlinked when GC runs by maintaining a
+      # reference to them.
```

Comment:
Can this really happen, that a tempfile is removed suddenly while you're using it? I would remove all these hacks from the `diff` gem itself, and run its test suite to see if something breaks.

Reply:
I think I would prefer an implementation that holds references to the tempfiles until the diff is generated, prints the diff, and then lets the tempfiles get deleted as part of garbage collection when the process is shutting down. It feels like a waste of code and effort to explicitly unlink the tempfiles ourselves when Ruby is about to do it for us when the process ends.